Add Kubernetes IPAM mode

tgraf opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/10407)
<!-- Reviewable:end -->

aanm commented on Documentation/concepts/ipam/kubernetes.rst:
```diff
@@ -0,0 +1,43 @@
+	  ``--allocate-node-cidrs`` flag to indicate to Kubernetes that PodCIDR
+	  ranges should be allocated.
+
+**via v1.Node annotation**
+
+=================================== ============================================================
+Annotation                          Description
+=================================== ============================================================
+``io.cilium.network.ipv4-pod-cidr`` IPv4 PodCIDR range
+``io.cilium.network.ipv6-pod-cidr`` IPv6 PodCIDR range
+=================================== ============================================================
+
+.. note:: The annotation-based mechanism is primiarly useful in combination with
+	  older Kubernetes versions which do not support ``Spec.PodCIDRs`` yet
+	  but support for both IPv4 and IPv6 is enabled.
```
Why are we adding this if we are removing the node annotation mode in 1.8? Will this still work?

tgraf replied:
Yes, the annotations can still be set manually by the user. I think it's a good fallback way for k8s versions which do not supports `PodCIDRs`. Thoughts?

aanm replied:
But how? we don't have a node watcher in cilium-agent AFAIK.

tgraf replied:
`waitForNodeInformation()` will `GET` the Kubernetes node and wait for the PodCIDR to be populated by either `PodCIDR`, `PodCIDRs` field or annotations.

aanm replied:
oh that makes more sense.

aanm reviewed (changes requested):
Also, shouldn't we add something in the upgrade guide with regarding the default value of the `ipam` option?

tgraf committed:
ipam: Remove direct access to option.Config

In order to test different IPAM modes easily in the future, remove direct
access to option.Config in order to allow mocking the configuration interface.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
ipam: Explicitly set the legacy hostscope IPAM as default

Explicitly set the current default IPAM mode in the viper option. This enables
introduction of a better default.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
ipam: Add "kubernetes" IPAM mode

This documents an IPAM mode which was already supported and widely used but
difficult to discover. The new "kubernete" IPAM mode requires the PodCIDR for
all enabled address families to be provided via the k8s Node resource and then
uses the standard hostscope allocator.

Signed-off-by: Thomas Graf <thomas@cilium.io>
